druid/src/widget/virt_list.rs

```diff
@@ -0,0 +1,401 @@
+                }
+            }
+            _ => {
+                data.for_in_mut(self.visible_data_range(data), |child_data, i| {
+                    self.child(i).event(ctx, event, child_data, env);
+                });
+            }
+        }
+    }
+
+    fn lifecycle(&mut self, ctx: &mut LifeCycleCtx, event: &LifeCycle, data: &C, env: &Env) {
+        if let LifeCycle::WidgetAdded = &event {
+            if self.update_child_count(data) {
+                ctx.children_changed();
+            }
+        }
+        data.for_in(self.visible_data_range(data), |child_data, i| {
+            self.child(i).lifecycle(ctx, event, child_data, env);
+        });
+    }
+
+    fn update(&mut self, ctx: &mut UpdateCtx, old_data: &C, data: &C, env: &Env) {
+        data.for_in(self.visible_data_range(data), |child_data, i| {
+            self.child(i).update(ctx, child_data, env);
+        });
+        if self.update_child_count(data) {
+            ctx.children_changed();
+        } else if old_data.data_len() != data.data_len() {
+            // The number of visible children haven't changed, but the total number
+            // of rows did, and therefore our total size.  We need to request layout.
+            ctx.request_layout();
+        }
+    }
+
+    fn layout(&mut self, ctx: &mut LayoutCtx, bc: &BoxConstraints, data: &C, env: &Env) -> Size {
+        let range = self.visible_data_range(data);
+        let height_above = self.child_size.height * range.start as f64;
+        let height_below = self.child_size.height * (data.data_len() - range.end) as f64;
+
+        let mut paint_rect = Rect::ZERO;
+        let mut width = bc.min().width;
+        let mut y = 0.0;
+
+        y += height_above;
+        data.for_in(range, |child_data, i| {
+            let child = self.child(i);
+            let child_bc = BoxConstraints::new(
+                Size::new(bc.min().width, 0.0),
+                Size::new(bc.max().width, std::f64::INFINITY),
+            );
+            let child_size = child.layout(ctx, &child_bc, child_data, env);
+            let rect = Rect::from_origin_size(Point::new(0.0, y), child_size);
+            child.set_layout_rect(ctx, child_data, env, rect);
+            paint_rect = paint_rect.union(child.paint_rect());
+            width = width.max(child_size.width);
+            y += child_size.height;
+        });
+        y += height_below;
+
+        let my_size = bc.constrain(Size::new(width, y));
+        let insets = paint_rect - my_size.to_rect();
+        ctx.set_paint_insets(insets);
+        my_size
+    }
+
+    fn paint(&mut self, ctx: &mut PaintCtx, data: &C, env: &Env) {
+        data.for_in(self.visible_data_range(data), |child_data, i| {
+            self.child(i).paint(ctx, child_data, env);
+        });
+    }
+}
+
+/// This iterator-like trait enables using `VirtList` with any kind of `Data`,
+/// with efficient iteration of sub-ranges.
+pub trait RangeIter<T>: Data {
```

Comment:
I'm not quite sure why you want ranges here, rather than direct indexed access from outside. 
Its similar to my IndexedItems in druid_table.  
If you think about sorting you can see why random access is the expected case. 
https://github.com/rjwittams/druid_table/blob/123b2d9ca8a446a551b6c2edf87c87075dea5bce/src/data.rs#L12

Also associated type for the iterated item (T) should be preferred - as you are not going to implement RangeIter for different T on the same type it will make complex inference situations simpler. (Worth looking round the Rust library types on this kind of thing).

Reply:
> I'm not quite sure why you want ranges here, rather than direct indexed access from outside.

Well, the access is sequential, `narrow()`ing `im::Vector` is faster than direct `.get()`, and the index irrelevant, so this seemed natural.

> as you are not going to implement RangeIter for different T on the same type

Actually I do that, but I agree it probably wasn't the intended case.